backend/src/scripts/utility/teamspaces/removeIncompleteRevisions.js

```diff
@@ -0,0 +1,186 @@
+	if (refAttribute) {
+		const projection = {};
+		projection[refAttribute] = 1;
+
+		const filesFilter = { ...filter };
+		filesFilter[refAttribute] = { $exists: true };
+
+		const results = await find(teamspace, collection, filesFilter, projection);
+		const filenames = [];
+
+		for (const record of results) {
+			const fileRefs = record[refAttribute];
+			if (fileRefs) {
+				if (TypeChecker.isString(fileRefs)) {
+					filenames.push(fileRefs);
+				} else {
+					for (const entry of Object.values(fileRefs)) {
+						filenames.push(entry);
+					}
+				}
+			}
+		}
+
+		await processFilesAndRefs(teamspace, collection, filenames);
+	}
+
+	await deleteMany(teamspace, collection, filter);
+};
+
+const processModelStash = async (teamspace, model, revId) => {
+	await processFilesAndRefs(
+		teamspace,
+		`${model}.stash.json_mpc`,
+		undefined,
+		{ filename: { $regex: `^/${teamspace}/${model}/revision/${UUIDToString(revId)}/` } },
+	);
+
+	await processFilesAndRefs(
+		teamspace,
+		`${model}.stash.json_mpc.ref`,
+		undefined,
+		{ _id: { $regex: `^${UUIDToString(revId)}/` } },
+	);
+
+	const unity3d = await find(teamspace, `${model}.stash.unity3d`, { _id: revId }, { assets: 1, jsonFiles: 1 });
+	for (const { assets, jsonFiles } of unity3d) {
+		// eslint-disable-next-line no-await-in-loop
+		await processFilesAndRefs(teamspace, `${model}.stash.unity3d.ref`, assets.map((filename) => filename.replace(`/${teamspace}/${model}/`, '')));
+		// eslint-disable-next-line no-await-in-loop
+		await processFilesAndRefs(teamspace, `${model}.stash.json_mpc.ref`, jsonFiles.map((filename) => filename.replace(`/${teamspace}/${model}/`, '')));
+		// eslint-disable-next-line no-await-in-loop
+		await processFilesAndRefs(teamspace, `${model}.stash.json_mpc.ref`, assets.map((filename) => filename.replace(`/${teamspace}/${model}/`, '').replace('unity3d', 'json.mpc')));
+		// eslint-disable-next-line no-await-in-loop
+		await processFilesAndRefs(teamspace, `${model}.stash.src.ref`, assets.map((filename) => filename.replace(`/${teamspace}/${model}/`, '').replace('unity3d', 'src.mpc')));
+	}
+
+	await processCollection(teamspace, `${model}.stash.unity3d`, { _id: revId }, '_extRef');
+	await processCollection(teamspace, `${model}.stash.3drepo`, { rev_id: revId }, '_extRef');
+};
+
+const processModelSequences = async (teamspace, model, revId) => {
+	const sequences = await find(teamspace, `${model}.sequences`, { rev_id: revId }, { frames: 1 });
+	const sequenceFilenames = [];
+
+	for (const { _id, frames } of sequences) {
+		if (frames) {
+			for (const { state } of Object.values(frames)) {
+				sequenceFilenames.push(state);
+				// eslint-disable-next-line no-await-in-loop
+				await processCollection(teamspace, `${model}.activities`, { sequenceId: _id }, '_extRef');
+				// eslint-disable-next-line no-await-in-loop
+				await processFilesAndRefs(teamspace, `${model}.activities.ref`, undefined, { _id: UUIDToString(_id) });
```

Comment:
Not that it makes any difference, but isn't this just filename?

I think ti's a bit confusing to have both filename and filters, just have filters?

Reply:
Changed to just filters. The `filenames` was there as `GridFS.removeFiles()` takes a list of filenames but I guess `FilesManager.removeFilesWithMeta()` should also handle GridFS.